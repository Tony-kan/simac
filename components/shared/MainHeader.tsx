"use client";

import {
  mainNavLinks,
  subNavLinks,
  quickLinksDropdown,
} from "@/constants/links";
import React, { useState, useEffect } from "react";
import { Search, Menu, X, ChevronDown } from "lucide-react";
import Link from "next/link";
import { motion, AnimatePresence } from "framer-motion";
import Image from "next/image";
import { Simac_Logo } from "@/constants/images";

//Todo: create a mobile_tablet nav,desktop nav
//Todo: make it responsive - mobile first. Add smooth framer animations/transitions
//Todo: Add search functionality.Note: quickLinks(dropdown menu),expand when hover - mobile when clicked
//! Alert,Do not use this function
//? questions need attentions
//* Highlighted comments */

// --- Animation Variants ---
// Mobile Menu: Main container slide-in
const mobileMenuVariants = {
  closed: { x: "100%" },
  open: {
    x: "0%",
    transition: { type: "tween", ease: "easeInOut", duration: 0.4 },
  },
};

// Mobile Menu: Staggered links container
const mobileLinkContainerVariants = {
  closed: { transition: { staggerChildren: 0.05, staggerDirection: -1 } },
  open: { transition: { staggerChildren: 0.07, delayChildren: 0.3 } },
};

// Mobile Menu: Individual link animation
const mobileLinkVariants = {
  closed: { opacity: 0, x: 50 },
  open: { opacity: 1, x: 0 },
};

// Mobile Menu: Collapsible section animation
const collapsibleVariants = {
  open: {
    opacity: 1,
    height: "auto",
    transition: { duration: 0.3, ease: "easeInOut" },
  },
  closed: {
    opacity: 0,
    height: 0,
    transition: { duration: 0.3, ease: "easeInOut" },
  },
};

// Desktop: Quick Links dropdown
const quickLinksVariants = {
  hidden: { opacity: 0, y: -10 },
  visible: { opacity: 1, y: 0, transition: { duration: 0.2 } },
};

// Header: Nav/Search switch
const navSearchVariants = {
  initial: { opacity: 0 },
  animate: { opacity: 1, transition: { duration: 0.3 } },
  exit: { opacity: 0, transition: { duration: 0.2 } },
};

/**
 * The main header component for the website.
 *
 * This component renders the main header for the website, including the logo,
 * navigation links, and search bar. It also handles the mobile menu overlay.
 *
 * The component uses the `useState` and `useEffect` hooks to manage the state of
 * the mobile menu and the search bar.
 *
 * The component renders a `header` element with a `nav` element inside it. The
 * `nav` element contains the main navigation links and the search bar. The
 * component also renders a mobile menu overlay that is displayed when the user
 * clicks on the mobile menu button.
 *
 * @return {React.ReactElement} The main header component.
 */
const MainHeader = () => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const [isSearchOpen, setIsSearchOpen] = useState(false);
  const [isQuickLinksOpen, setIsQuickLinksOpen] = useState(false);
  const [isMobileQuickLinksOpen, setIsMobileQuickLinksOpen] = useState(false);

  // Effect to handle body scroll lock for mobile menu
  useEffect(() => {
    document.body.style.overflow = isMobileMenuOpen ? "hidden" : "auto";
    return () => {
      document.body.style.overflow = "auto";
    };
  }, [isMobileMenuOpen]);

  useEffect(() => {
    if (!isMobileMenuOpen) {
      setIsMobileQuickLinksOpen(false);
    }
  }, [isMobileMenuOpen]);

  /**
   * A functional component that renders a styled link using the Next.js Link component.
   *
   * @param {string} href - The URL the link points to.
   * @param {React.ReactNode} children - The content to be displayed inside the link.
   * @param {string} [className] - Optional additional CSS classes for styling.
   * @param {React.MouseEventHandler<HTMLAnchorElement>} [onClick] - Optional click handler function.
   */

  const NavLink = ({
    href,
    children,
    className = "",
    onClick = () => {},
  }: INavLinkProps) => (
    <Link
      href={href}
      className={`text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 ${className}`}
      onClick={onClick}
    >
      {children}
    </Link>
  );

  return (
    <>
      <header className="relative border-b border-gray-300  bg-white h-[12vh] min-h-[80px] lg:h-[18vh] lg:min-h-[120px] shadow-sm">
        {/* Desktop header  */}
        <div className="container px-6 h-full max-w-screen flex justify-between items-center gap-4">
          <div className="flex-shrink-0 ml-0 lg:ml-8 size-18 md:size-22 lg:size-32">
            <Link
              href="/"
              className="text-2xl lg:text-3xl font-bold text-gray-800"
            >
              <Image
                src={Simac_Logo}
                alt="Logo"
                width={40}
                height={40}
                className="w-auto h-auto"
              />
            </Link>
          </div>
          <div className="flex-grow flex justify-end h-full">
            <AnimatePresence mode="wait">
              {!isSearchOpen ? (
                <motion.div
                  key="nav-view"
                  variants={navSearchVariants}
                  initial="initial"
                  animate="animate"
                  exit="exit"
                  className="w-full h-full"
                >
                  <div className="hidden lg:flex flex-col justify-between h-full mr-4">
                    <div className="flex justify-end items-center gap-6 mt-8 text-sm text-gray-500">
                      {subNavLinks.map((item) =>
                        item.title === "Quick Links" ? (
                          <motion.div
                            key={item.id}
                            className="relative"
                            onHoverStart={() => setIsQuickLinksOpen(true)}
                            onHoverEnd={() => setIsQuickLinksOpen(false)}
                          >
                            <button className="flex items-center gap-1 hover:text-[#5C1B23]">
                              {item.title}
                              <motion.div
                                animate={{ rotate: isQuickLinksOpen ? 180 : 0 }}
                              >
                                <ChevronDown size={16} />
                              </motion.div>
                            </button>
                            <AnimatePresence>
                              {isQuickLinksOpen && (
                                <motion.div
                                  variants={quickLinksVariants}
                                  initial="hidden"
                                  animate="visible"
                                  exit="hidden"
                                  className="absolute top-full left-1/2 -translate-x-1/2 mt-2 w-48 bg-white rounded-md shadow-lg border z-20"
                                >
                                  {quickLinksDropdown.map((link) => (
                                    <NavLink
                                      key={link.id}
                                      href={link.url}
                                      className="block w-full text-left px-4 py-2 text-sm hover:bg-gray-100"
                                    >
                                      {link.title}
                                    </NavLink>
                                  ))}
                                </motion.div>
                              )}
                            </AnimatePresence>
                          </motion.div>
                        ) : (
                          <NavLink key={item.id} href={item.url}>
                            {item.title}
                          </NavLink>
                        )
                      )}
                      <button
                        onClick={() => setIsSearchOpen(true)}
                        aria-label="Open search"
                        className="hover:text-[#5C1B23]"
                      >
                        <Search size={18} />
                      </button>
                    </div>
                    <nav className="flex justify-end items-center gap-6 mb-4 font-medium">
                      {mainNavLinks.map((item) => (
                        <NavLink
                          key={item.id}
                          href={item.url}
                          className="text-base font-semibold"
                        >
                          {item.title}
                        </NavLink>
                      ))}
                    </nav>
                  </div>
                  <div className="lg:hidden flex items-center justify-end h-full gap-4">
                    <button
                      onClick={() => setIsSearchOpen(true)}
                      aria-label="Open search"
                      className="text-gray-600 hover:text-[#5C1B23]"
                    >
                      <Search size={24} />
                    </button>
                    <button
                      onClick={() => setIsMobileMenuOpen(true)}
                      aria-label="Open menu"
                      className="text-gray-600 hover:text-[#5C1B23]"
                    >
                      <Menu size={28} />
                    </button>
                  </div>
                </motion.div>
              ) : (
                <motion.div
                  key="search-view"
                  variants={navSearchVariants}
                  initial="initial"
                  animate="animate"
                  exit="exit"
                  className="w-full h-full flex items-center"
                >
                  <form className="w-full relative">
                    <input
                      type="search"
                      placeholder="Search and press enter..."
                      className="w-full pl-4 pr-12 py-2 border-2 border-gray-300 focus:border-blue-500 focus:outline-none rounded-md text-base lg:text-lg transition-colors"
                      autoFocus
                    />
                    <div className="absolute top-1/2 right-4 -translate-y-1/2 flex items-center gap-2">
                      <Search size={20} className="text-gray-400" />
                      <button
                        onClick={() => setIsSearchOpen(false)}
                        type="button"
                        aria-label="Close search"
                        className="text-gray-500 hover:text-gray-800"
                      >
                        <X size={24} />
                      </button>
                    </div>
                  </form>
                </motion.div>
              )}
            </AnimatePresence>
          </div>
        </div>
      </header>

      {/* --- Mobile Menu View--- */}
      <AnimatePresence>
        {isMobileMenuOpen && (
          <motion.div
            variants={mobileMenuVariants}
            initial="closed"
            animate="open"
            exit="closed"
            className="fixed inset-0 bg-white z-50 px-8 lg:hidden"
          >
            <div className="flex justify-between items-center p-4 border-b">
              <h2 className="font-bold text-xl">Menu</h2>
              <button
                onClick={() => setIsMobileMenuOpen(false)}
                aria-label="Close menu"
              >
                <X size={28} />
              </button>
            </div>
            <motion.nav
              variants={mobileLinkContainerVariants}
              initial="closed"
              animate="open"
              className="flex flex-col p-4 space-y-2"
            >
              {/* Main Links */}
              {mainNavLinks.map((item) => (
                <motion.div key={item.id} variants={mobileLinkVariants}>
                  <NavLink
                    onClick={() => setIsMobileMenuOpen(false)}
                    href={item.url}
                    className="text-lg py-2 block w-full font-semibold"
                  >
                    {item.title}
                  </NavLink>
                </motion.div>
              ))}

              <div className="border-t my-4"></div>

              {/* Sub Links with Collapsible "Quick Links" */}
              {subNavLinks.map((item) => (
                <motion.div
                  key={item.id}
                  variants={mobileLinkVariants}
                  className="overflow-hidden"
                >
                  {item.title === "Quick Links" ? (
                    <>
                      <button
                        onClick={() =>
                          setIsMobileQuickLinksOpen(!isMobileQuickLinksOpen)
                        }
                        className="w-full flex justify-between items-center text-base font-semibold text-gray-500 py-1 hover:text-[#5C1B23] transition-colors"
                      >
                        <span>{item.title}</span>
                        <motion.div
                          animate={{ rotate: isMobileQuickLinksOpen ? 180 : 0 }}
                        >
                          <ChevronDown size={20} />
                        </motion.div>
                      </button>
                      <AnimatePresence>
                        {isMobileQuickLinksOpen && (
                          <motion.div
                            variants={collapsibleVariants}
                            initial="closed"
                            animate="open"
                            exit="closed"
                            className="flex flex-col pl-4 mt-1"
                          >
                            {quickLinksDropdown.map((link) => (
                              <NavLink
                                key={link.id}
                                href={link.url}
                                onClick={() => setIsMobileMenuOpen(false)}
                                className="py-2 text-sm block w-full"
                              >
                                {link.title}
                              </NavLink>
                            ))}
                          </motion.div>
                        )}
                      </AnimatePresence>
                    </>
                  ) : (
                    <NavLink
                      onClick={() => setIsMobileMenuOpen(false)}
                      href={item.url}
                      className="text-base text-gray-500 py-1 block w-full font-semibold"
                    >
                      {item.title}
                    </NavLink>
                  )}
                </motion.div>
              ))}
            </motion.nav>
          </motion.div>
        )}
      </AnimatePresence>
    </>
  );
};

export default MainHeader;
