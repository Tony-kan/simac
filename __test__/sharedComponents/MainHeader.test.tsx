import React from "react";
// import { render, fireEvent, waitFor } from "@testing-library/react";
import { render } from "@testing-library/react";

import MainHeader from "@/components/shared/MainHeader";

describe("MainHeader component", () => {
  it("renders correctly", () => {
    const { container } = render(<MainHeader />);
    expect(container).toMatchSnapshot();
  });

  //   it("opens and closes mobile menu", () => {
  //     const { getByRole, getByText } = render(<MainHeader />);
  //     const menuButton = getByRole("button", { name: "Open menu" });
  //     fireEvent.click(menuButton);
  //     expect(getByText("Menu")).toBeInTheDocument();
  //     fireEvent.click(getByRole("button", { name: "Close menu" }));
  //     expect(getByText("Menu")).not.toBeInTheDocument();
  //   });

  //   it("opens and closes search bar", () => {
  //     const { getByRole, getByText } = render(<MainHeader />);
  //     const searchButton = getByRole("button", { name: "Open search" });
  //     fireEvent.click(searchButton);
  //     expect(getByText("Search and press enter...")).toBeInTheDocument();
  //     fireEvent.click(getByRole("button", { name: "Close search" }));
  //     expect(getByText("Search and press enter...")).not.toBeInTheDocument();
  //   });

  //   it("opens and closes quick links dropdown", () => {
  //     const { getByText, getByRole } = render(<MainHeader />);
  //     const quickLinksButton = getByRole("button", { name: "Quick Links" });
  //     fireEvent.mouseOver(quickLinksButton);
  //     expect(getByText("Quick Links")).toBeInTheDocument();
  //     fireEvent.mouseOut(quickLinksButton);
  //     expect(getByText("Quick Links")).not.toBeInTheDocument();
  //   });

  //   it("opens and closes mobile quick links dropdown", () => {
  //     const { getByText, getByRole } = render(<MainHeader />);
  //     const mobileMenuButton = getByRole("button", { name: "Open menu" });
  //     fireEvent.click(mobileMenuButton);
  //     const mobileQuickLinksButton = getByRole("button", { name: "Quick Links" });
  //     fireEvent.click(mobileQuickLinksButton);
  //     expect(getByText("Quick Links")).toBeInTheDocument();
  //     fireEvent.click(mobileQuickLinksButton);
  //     expect(getByText("Quick Links")).not.toBeInTheDocument();
  //   });
});
