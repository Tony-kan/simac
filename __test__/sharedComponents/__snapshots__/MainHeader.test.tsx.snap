// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainHeader component renders correctly 1`] = `
<div>
  <header
    class="relative border-b border-gray-300  bg-white h-[12vh] min-h-[80px] lg:h-[18vh] lg:min-h-[120px] shadow-sm"
  >
    <div
      class="container px-6 h-full max-w-screen flex justify-between items-center gap-4"
    >
      <div
        class="flex-shrink-0 ml-4 lg:ml-8 size-18 md:size-22 lg:size-32"
      >
        <a
          class="text-2xl lg:text-3xl font-bold text-gray-800"
          href="/"
        >
          <img
            alt="Logo"
            class="w-auto h-auto"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </a>
      </div>
      <div
        class="flex-grow flex justify-end h-full"
      >
        <div
          class="w-full h-full"
          style="opacity: 0;"
        >
          <div
            class="hidden lg:flex flex-col justify-between h-full mr-4"
          >
            <div
              class="flex justify-end items-center gap-6 mt-8 text-sm text-gray-500"
            >
              <div
                class="relative"
              >
                <button
                  class="flex items-center gap-1 hover:text-[#5C1B23]"
                >
                  Quick Links
                  <div>
                    <svg
                      aria-hidden="true"
                      class="lucide lucide-chevron-down"
                      fill="none"
                      height="16"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m6 9 6 6 6-6"
                      />
                    </svg>
                  </div>
                </button>
              </div>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 "
                href="/donate"
              >
                Donate
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 "
                href="/store"
              >
                Store
              </a>
              <button
                aria-label="Open search"
                class="hover:text-[#5C1B23]"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-search"
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 21-4.34-4.34"
                  />
                  <circle
                    cx="11"
                    cy="11"
                    r="8"
                  />
                </svg>
              </button>
            </div>
            <nav
              class="flex justify-end items-center gap-6 mb-4 font-medium"
            >
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/about"
              >
                About
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/programs"
              >
                Programs
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/events"
              >
                News & Events
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/affiliates"
              >
                Affiliates
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/blog"
              >
                Blog
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/gallery"
              >
                Gallery
              </a>
              <a
                class="text-gray-700 hover:text-[#5C1B23] transition-colors duration-300 text-base"
                href="/contact-us"
              >
                Contact
              </a>
            </nav>
          </div>
          <div
            class="lg:hidden flex items-center justify-end h-full gap-4"
          >
            <button
              aria-label="Open search"
              class="text-gray-600 hover:text-[#5C1B23]"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-search"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m21 21-4.34-4.34"
                />
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
              </svg>
            </button>
            <button
              aria-label="Open menu"
              class="text-gray-600 hover:text-[#5C1B23]"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-menu"
                fill="none"
                height="28"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="28"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 12h16"
                />
                <path
                  d="M4 18h16"
                />
                <path
                  d="M4 6h16"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
